### PR TITLE
Improve errors for bind/discard 

### PR DESCRIPTION
**Pull request by paf31**
For #2732

### CODE REVIEW

**Comment by garyb** [comment]
Any thoughts on switching to `bind_` rather than `discard`?

**Comment by paf31** [comment]
I guess it would need a (technically) breaking change in Prelude, but I don't mind, we can.

**Comment by garyb** [comment]
Oh! Now you mention it, in the `ps-0.11` branch I made for the `Prelude`, I made another breaking change anyway - by removing the `EuclideanRing` and `Field` instances for `unit`. I doubt they're used much, if anywhere, but I would have completely forgotten about that if you hadn't said.

I'm fine with fixing everything again for that. Since everything from `-eff` down is "broken" anyway it's not much of a difference to update for `-prelude` too.

**Comment by paf31** [comment]
I was assuming we'd need to make some major releases anyway, so sure, let's change it.

**Comment by paf31** [comment]
This now includes fixes to the type checker (error hints were backwards when solving instances - this should improve other things too) and the parser to generally improve error quality in `do` blocks.

File: src/Language/PureScript/Errors.hs [hunk 1]
```diff
@@ -1120,7 +1120,7 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs) e = flip evalS
 
   -- | Simplify an error message
   simplifyErrorMessage :: ErrorMessage -> ErrorMessage
-  simplifyErrorMessage (ErrorMessage hints simple) = ErrorMessage (simplifyHints hints) simple
+  simplifyErrorMessage (ErrorMessage hints simple) = ErrorMessage (traceShow (filter ((PositionHint == ) . hintCategory) hints) $ simplifyHints hints) simple
```
Comment (garyb) [inline src/Language/PureScript/Errors.hs]: `traceShow`? :smile:

Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Oops :)

**Comment by paf31** [comment]
@garyb If this looks ok, I'd like to merge it and then fix the naming issue separately. Sound good?

**Pull request merged**